### ISSUE TITLE
Karaoke+Chords   and Music Notation+Tab

### ISSUE THREAD

**Issue by KaraJam-dev** [spessasus/SpessaSynth#86]
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. Ex. It would be cool to [...]

**Is your feature related to the sound library or the app/website?**
The library is responsible for making the sound and the website is the keyboard, renderer etc.

** My Project is called KaraJam(TM) I am a musician who wants to build a cool program to help musicians learn songs.
1. I am not a programmer
2. I have dyslexia and struggle to keep track of code, snippets etc drive me crazy.
3. I have been using Clause and ChatGTP to try and do this. It works then it doesn't work, ugh
4. This has been a dream of mine for many many years, I have tried all kinds of efforts including on Lynux using Jack and the many opensource libraries there. I gave up on Lynux as it required a better education than what I have. I like to say "Duh me play drums" 
5. The basic are to create a new pane that presents First Karaoke and Chords, for this I am trying to use the Fraigo Karaoke Parsing feature, for some reason the Shift+T feature does not read many many .mid and .kar files not sure why but Fraigo reads them all? 
6. I am trying to use Vexflow Libraries to parse chords and present them under the lyrics. This would be in a new pane like the main screen pane and the music player pane called "Karaoke+Chords".
7. I am also trying to use Vexflow Libraries to parse music notation and tab. The concept here is to have a musical note on the individual channels of the Synthesizer Controller, when you click that track's note button the music notation for that track is added to the Karaoke+Chords Screen/pane. On the Music Notation sub/pane there is a button for TAB, this will convert the Notation into either Guitar TAB or if it is on Track 10 it will be converted to Drum Tab. These of course are toggles.
8. I am trying to figure out how to make my new features in a new pane, they keep being added to the main screen or the music player. I tried duplicating the content for music_player_mode.js and creating the new Karaoke+chords but it is only adding the place marker to the bottom of both the main screen and the music player screen. 

I would gladdly pay you to help me build KaraJam(TM), I don't have lots of money but will try to raise any money required. I can raise money but to write code I am help less.

P.S GREAT JOB on this program, I can't begin to list all the cool features and how musicians can put this to use. I am an old rocker from the 70s-80s and these days I am out of it, but would love to start practicing songs again, just to keep up the skills, and why not share that with other musician's as well.

Sincerely, Pat
KaraJam(TM) 
a2dme@live.com

**Describe alternatives you've considered**

**Additional context**

**Comment by spessasus** [comment]
Hi,

I am not sure if I understood you correctly. Is your request related to your KaraJam program or SpessaSynth? If the latter, what would you like to get added into the program?

**Comment by KaraJam-dev** [comment]
I am working with ChatGTP to incorporate SpessaSynth Fraigo Karaoke and VexFlow.

Been dreaming of KaraJam(TM) for twenty years. Heres the basic layout as I see it.


  1.
Karaoke+Chords

Program Open in a Karaoke+Chords Mode, we get Karaoke by Fraigo because of it robust Lyrics Handling, SpessaSynth only handles one of the over 100 .mid and .kar files I have, to parse the lyrics, Fraigo can parse them all. Then we use Vexflow to parse the chords, when changing key the program must be able to transpose the chord changes.. So the main UI is Karaoke+Chords like a musician's cheatsheet.


  1.
Midi Mixer and Synth

Then we come to the Midi Mixer portion, this should look like a Normal Audio Mixer, with Channels and a Master Section, including Reverb, Chorus, EQ Hi Mid and Low the Channels will have the ability to change instruments (GM) and import individual SF2 instruments GM MIDI Guitar sounds really suck). I would separate your Synthesizer Controller into two parts, 1 Midi Mixer and 2 a Synthesizer section for Modulation other Synth related parameters. I would like to see the Synthesizer Section expanded to include the possibility of an actual Synthesizer segment looking similar to the old Mooge Synth, using FM wave Synthesis or other more modern paramemters.


  1.
Music Notation / TAB

When in the Midi Mixer portion we can hit a Musical Note at the top of the channel and the mixer goes away and Music Notation for that channel is added to the bottom of the Karaoke+Chords pane. On the Music Notation Pane is a TAB button that will convert that channel to Tableture for Guitar, any channel can be viewed in Guitar Tab accept for Channel ten. Channel Ten needs to have Drum TAB of course. I understand that VexFlow is ideal for this, but have not got there yet.


  1.
Mix Down "Save Audio"

The Master Section of the MIDI Mixer should have the ability to either import a Compressor/Limiter or have one resident. I have looked at VST as a good possibility for signal Processing, but still exploring the possibilities here. Midi files can be all over the place in channel levels and even the difference between files can be an issue, we want a FM Radio Style compression applied at the Master output.

SpessaSynth is a great foundation for my project, I am struggling to get what I want from ChatGTP, some of it's code works great some not so great, up and down up and down excitement disappointment. I am not a programmer and suffer from dyslexa so when I have to work with snippets of code I get lost quickly. I have had some great success with Performous, but when I saw SpessaSynth was web based and did not require installation and worked on all platforms with a browser, i fell in love with SpessaSynth.


  1.
 KaraJam(TM) the game

KaraJam(TM) the game, I envision going forward after establishing a user base for KaraJam(TM) that a game would be the next logical step. 1 like Roch Band or Guitar Hero, but for real musicians. 2. Midi input for polyphonic inputs and USB audio for Monophonic inputs. If you can imagine replacing the Keyboard with a Guitar Fret Board SpessaSynth already has the parts and pieces to make this happen.
Picture the singer get the note lines like in Performous, the Guitar player gets the vertical fret board (Non 3d) Key Boards well you got that beat already. It is possible (tested with Performous) to even do Brass and Woodwind (Monophonic) but the UI is still unclear.

To begin with I want the Karaoke+Chords working, then move to the MIDI Mixer and Synth, then add the Music Notation and TAB. Your Save Audio section has the potential to be great for not only the Mix Down but also for saving custom soundfonts once custom channel SF2 files have been added. There are some much better Guitar SF2 and VST instrument to be used out there and most of them are free, though I saw a few paid VST that might really work better for guitar.

There is still more to talk about here. By Linking the Music Notation portion to a network we could envision a School Band Teacher using the Software for his/her whole band. I did this with Lynux many years ago we used Muscore and by sending MIDI time code had them all synced up, though back then slower computers would have trouble with lag, I resolved that by only having ghe Start/stop Command heard on those slow machines, kind of worked but once in a while even then it would loose it's way. Now days it seems they are sending midi time code over the network much better, this opens the door to have Musician forming band from home, jamming etc, though even then inorder for them to hear the other guy we have to reduce the sample rate to almost being unworthy of the effort. The best way I see is to let each player just hear the Midi file and practice along, adding a video phone like feature that disables during play back has worked for me jamming with friends in the past, but again the audio sample rate had to be really bad and even then the latency got to be an issue. So far I only see practicing together and chatting later as a serious option. This is clear an expansion option but has the potential for a huge user base. Kids have become so anti-social these days.

So you would think that is it, keep track if you can, now when it comes to talking KaraJam(TM) to a venue, lets talk about what modern technology has open up for musician's and how MIDI could be used in a live performance use base. I played in Bands in the 70-80-90s and I use to blow peoples minds by preprogramming the lights using midi, it could be dead on the beat and have all kinds of effects, like Strobe or Specials doing cool effects right in time to the music. Spot lights would come on right when the singer would take the mike and using sysex I could make it so the guitar player never had to touch their pedals. I used a Midi Light Controller by Sun and a Sysex/Midi-Controller controller byBerranger (I think). By Using a click track in my ear (as the drummer) I cold keep the band in sync with the Midi File. Cool Right? The main issue was no more endless guitar solos and by using a FM style Compressor/Limiter on the mains, we had a Album Quality sound. Today with Amp Simulators and Electric Drums etc we could do a sound and stage design that meets even the Workman's Compensation Board standards for SPL levels. Bringing back the live band days.

Practice at home, play the game till your good enough to perform and form bands online, they meet for the first time at the gig potentially. Lol Single Acts and Duo's can create backing tracks for use at the gig as well. Even create custom sound fonts, it's all there if we can merge SpessaSynth, Fraigo Karaoke and Vexflow CHords and Charts.

Well I said I would pay you to build the KaraJam Software but understand I might be able to raise a few grand 1-3 but I am not rich and once it's done it would still be open source and free for the users. Only once karajam.org and kchord.org are live and build a user base will the potential for the game and the venue system to explode on the scene.

For now the focus is to build Karajam(TM) on karajam.org. I already have the hosting package with Network Solutions ready to go. So the job if you can do it for a few grand is build the Karaoke+Chords, add the Midi Mixer/Synth add the Music Notation/TAB and host it on my GitHub server, KaraJam-Dev.

Once it is up and running, the next step is to build a user base that loves it and uses it all the time, then I can move toward the Proprietary Game and Venue systems.

Let me know what you think, till then I will continue pluggin away at SpessaSynth/Fraigo-karaoke/Vexflow dreaming of the day it all works and I can start both telling my friends to try it out and using it my self to learn the oldies. I sing and play both Leys and Guitar a bit, so that would be fun in and of it's self.

Complements on SpessaSynth great software wonderful features and wow so close to my dream for KaraJam(TM) by the way I have owned KaraJam the Trade Mark since 15 good till 2030 and will renew it then. So far it is only in Canada, but they tell me there are treaties that cover me in most countries. As soon as I can afford it I will expand that as well, and try to get into Music Licensing. I would love to create KaraJam(TM) Creator software around my sound and stage design that would be like having a robot band..lol Customized MIDI files with Lights SYSEX etc, would be cool but my lawyers says music licensing is big money to get involved in. Ugh.. For now there are tonnes of free Midi files and lots of places to buy them too.

Thanks for getting back to me, see the vision believe in the dream, lets make KaraJam(TM) together. Oh and revenue could be generated off of KaraJam.org via advertising and I would gladely share some of that income with you as well, if we had an agreement in principle where you are the software author and I am the Producer Founder. Regardless I am never giving up on my dream.

Pat Hobbis
KaraJam(TM)
________________________________

**Comment by KaraJam-dev** [comment]
To be clear I am using SpessaSynth as the foundation and adding libraries from Fraigo-karaoke and Vexflow to add features, is this ok? Chat GTP said your software was the best candidate for the new app.

I hope so i have invested many hours on this. It's a MIT license right?

Oh my musician buddies think I should not use KaraJam because it is associated with Karaoke, WHICH THEY HATE, I am now thinking I will release the software as KJam. So kjam.org will be the website.

Let me know what you think. The money is there if you want it, but I will keep banging away at the AI for now till I hear from you.

Pat
KaraJam(TM)
________________________________

**Comment by spessasus** [comment]
Hi,

1. First of all, you mentioned that this "Fraigo Karaoke" can parse some of your MIDI files while Spessasynth can't. Can you either upload them here or e-mail them to me? (spesekspesek@gmail.com) I'll fix up spessa's parser to make them work and then Fraigo won't be necessary.

2. You mentioned the quality of the built-in samples. I understand, the built-in soundfont is intended to be a small soundfont that gets loaded quickly. I wish I could make a different one the default, but nobody wants to download a 1GB file on the start.
I personally use a different, higher quality soundfont: this one (https://musical-artifacts.com/artifacts/1176) to be precise. It has some very high quality samples that will satisfy your needs. And for different SF2 files, sure. SpessaSynth supports SF list.

3. Musical notation is a topic that has been discussed in spessasynth before. I am considering implementing it, though doing it with Vexflow is out of the question. However, I am willing to help you integrate Vexflow into your `KaraJam/KJam` project, though I'd prefer us conversing over e-mail for that, as it is not related to SpessaSynth. Again, feel free to send me an e-mail if you're interested.

4. About some of the features you mentioned:
- MIDI ports are already supported. Both playing to other MIDI devices or MIDI input. Navigate to settings and they are available under the section named "MIDI Settings." Note that you have to give the permission for the website to access the MIDI devices and **restart the browser after you've connected any for them to show up!**
- VSTis simply don't work on the web as they are intended to be installed as sort of "drivers". So I can't help you with that one.
- FM Modulation is out of scope for SpessaSynth. Spessa is a SoundFont (or DLS) synthesizer, I.E. a wavetable synthesizer, or in other words, a fancy sampler. I can't help you with that one, though we could emulate it by using samples as well.
- Equalizer does sound quite easy to implement, by using `BiquadFilterNode` in the WebAudio API, so I can try to help you with that. 

5. About AI: I personally can't really recommend chat GPT or bing copilot for the task. GitHub Wikis (where SpessaSynth's documentation is hosted) aren't indexed by search engines and also by these assistants. The only AI i'd recommend for the task is GitHub Copilot, if you have access to that. Though, again, I am willing to help you with your project.

6. It is absolutely okay for you to use my project like that. In fact, i'm quite happy that it can help make your dream come true :)

And final question: Do you want me to add these features to spessasynth, or is it strictly related to `KaraJam`? If the latter, please reach out to me via e-mail and we'll continue from there.

Regards,
spessasus
(spesekspesek@gmail.com)

**Comment by KaraJam-dev** [comment]
I can't send you midi files it would be an infringement but you can simply search for free midi karaoke files there are lost of them out there not sure fraigo karaoke handles the so well? Spessasynth says no lyrics found to most files. Not sure why though.

Yes I am planning to release KJam.org in the new year. About Vexflow I am not stuck on any particular library. They said my Muscore score has the same capabilities. How ever it needs to be compiled first.? There are others as well but my research says that Vexflow will parse chords nicely like lyrics in sync. Have not tried it yet.

Just think about a young musician learn to play classic rock. We first want Karaoke plus chords like a musician cheat sheet then we want individual tracks to have either notation or TAB so they can slow it down and practice it a bit at a time.

KJam is intended to help young musicians learn songs in everyway possible.

Now as I said if I can pay you to do the programming I would be very happy because I am 63 with dyslexia and have never been able to get this done. You let me know what you want I will raise more money if I need to.

The other stuff is all just dreaming for now but if I can get a strong user base advertisers and subscription may open the door to revenue potential.

I really like the keyboard interface and we under how hard it would be to put a static guitar neck on the screen as an option. More can complex because guitar has the same note up and down the neck so how do we know which one the artist played.

For now karaoke plus chords is the starting ground if you would take that on. Let me know how much you want. Changing the look of the synthesizer controller to more of a traditional mixer would also be cool even without the EQ . You tell me what you want to get started. I will find a way. I can etransfer a deposit asap.

Great to hear from you.
Pat
KaraJam (TM)


Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by KaraJam-dev** [comment]
I was trying to set up the paypal account, it says I already have an account, but the banking information on that account is wrong, a really old account, so now I have to go to my bank and set this up again.

Darn I am frustrated, because there is going to be more money coming for you as we progress.

Got to get this sorted out today.

Oh also when I put your information into GitHub it can't find you?

Pat
KaraJam(TM)
